### PR TITLE
Human Vs Unbeatable Computer

### PR DESCRIPTION
**Pull request by sethpthomas91**

### CODE REVIEW

File: lib/UI.rb [hunk 1]
```diff
@@ -122,11 +112,18 @@ def welcome_user
   end
 
   def computer_thinking_message
-    puts "Thinking..."
+    thinking_messages = [
```
Comment (martingaston) [inline lib/UI.rb:115]: this is a nice touch for your design/UX 🎉

File: lib/computer_player.rb [hunk 2]
```diff
@@ -1,17 +1,82 @@
...
+    if game_logic.return_winner_symbol(symbol_arr) == minimizing_player.marker
+      return -10
+    elsif game_logic.return_winner_symbol(symbol_arr) == maximizing_player.marker
+      return 10
+    elsif game_logic.draw?(symbol_arr)
+      return 0
+    end
+
+    if is_maximizing
+      best_score = -100
```
Comment (martingaston) [inline lib/computer_player.rb]: major nitpick, this works absolutely fine with your algorithm, but `-Float::INFINITY` might the most arithmetically accurate starting number 😄

File: lib/UI.rb [hunk 3]
```diff
@@ -1,36 +1,33 @@
...
-    while true
-      user_input = gets.chomp
-      user_integer = input_to_int(user_input)
-      break if valid_input(user_integer)
-
-      prompt_invalid_input
-    end
-    user_integer
+  def get_input
+    user_input = gets.chomp
```
Comment (martingaston) [inline lib/UI.rb]: nitpick, don't think the variable here gets used 👍

File: lib/game_setup.rb [hunk 4]
```diff
@@ -5,26 +5,29 @@
...
     @game = nil
   end
 
   def main_menu
+    ui.clear_terminal_screen
     ui.welcome_message
     loop do
       ui.game_selection_prompt
-      user_input = ui.get_menu_selection
+      user_input = ui.return_choice_integer_between(1, 4)
```
Comment (martingaston) [inline lib/game_setup.rb]: i had to stare at this naming for a bit - wonder if using `get`/`gets` in the naming will help it sit more idiomatically in the rubyverse. something like `get_integer_between(1, 4)`

File: spec/ui_spec.rb [hunk 5]
```diff
@@ -19,31 +19,19 @@
...
-    end
-  end
 
-  describe '#get_human_play_first_input' do
-    it 'should accept 1 as a valid input' do
-      ui.stub(gets: '2')
-      expect(ui.get_human_play_first_input).to eq(2)
+    context 'user inputs 1' do
+      it 'should not return an integer of 2' do
+        expect_any_instance_of(UI).to receive(:get_input) { 1 }
```
Comment (martingaston) [inline spec/ui_spec.rb:32]: feels a bit like you're testing the double/implementation, here. 

what about the behaviour? `it 'will return the first valid integer within the target range'` and your double returns `{ 0, -1, "one", "1", 5.55, 10, nil, 1, 2 }` - then if you're asserting that you get `1` returned, you can assume that you've gone through the loop until you get valid input.

File: lib/game_setup.rb [hunk 6]
```diff
@@ -5,26 +5,29 @@
...
       case user_input
       when 1
         human_vs_human_builder
       when 2
         human_vs_computer
       when 3
+        computer_vs_computer_builder
```
Comment (martingaston) [inline lib/game_setup.rb:29]: nice, I like extracting the builders into their own methods

File: lib/game.rb [hunk 7]
```diff
@@ -20,39 +18,29 @@ def check_is_valid_human_move(user_input)
...
   end
 
-  def turn(player)
-    ui.turn_start(board)
+  def turn(board:, player:)
+    ui.turn_start(board, player)
     if player.is_computer == true
-      user_input = player.random_move
-      user_input = check_is_valid_computer_move(user_input)
+      user_input = player.move(maximizing_player: player, board: board)
```
Comment (martingaston) [inline lib/game.rb:40]: `maximizing_player` suggests your internals for the _unbeatable_ computer are leaking into your interface - presumably this isn't required for your non-unbeatable computer?

how might you lose the conditional and have a API where you can have something like, `player.move(board)` so you can use the open/closed principle?

File: spec/computer_player_spec.rb [hunk 8]
```diff
@@ -20,4 +22,23 @@
       expect(computer_player.assign_marker_as('X')).to eq('X')
     end
   end
+
+  describe '#make_unbeatable' do
```
Comment (martingaston) [inline spec/computer_player_spec.rb:26]: i think a good test case you might be missing: what would happen if two unbeatable computers were to play one another?

**Review by martingaston — COMMENTED** [review]
great job 🎉 can't believe you've got minimax up and running so quickly!  

one thing is that, well, this is a pretty big PR - there's quite a bit in here that doesn't relate to actually adding your minimax computer, but is modifying your design a little so you can add said feature. that might be a nice seam to split into a couple of PRs.

sometimes your fellow team can be somewhat time poor, so it can be helpful to remove as many possible barriers from them reviewing your PR - and the smaller the PR is, the fewer excuses they can find to actually take a look at it 😄

File: lib/UI.rb (hunk 1 above)
Comment (plopeziv) [reply lib/UI.rb:115]: +1 I think there was a hmmm in there somewhere that I also enjoyed

File: lib/computer_player.rb (hunk 2 above)
Comment (sethpthomas91) [reply lib/computer_player.rb]: Understood. This makes sense so it can be more applicable for other applications.

File: lib/game_setup.rb (hunk 4 above)
Comment (sethpthomas91) [reply lib/game_setup.rb]: Ruby doesn't seem to like the get wording at the beginning of methods. At least rubocop flags it as a no-no. I've been trying to go back and decrease those occurrences in my code.

File: lib/game.rb (hunk 7 above)
Comment (sethpthomas91) [reply lib/game.rb:40]: I talked with Pedro about this. I think this is a great idea that would decouple the functions and make the interface much better.

File: lib/UI.rb [hunk 9]
```diff
@@ -88,12 +78,12 @@ def prompt_draw
...
-  def prompt_invalid_space
-    puts 'Please enter a valid space'
+  def prompt_invalid_input(lower_int, upper_int)
+    puts "Please enter a valid selection between #{lower_int} and #{upper_int}"
   end
 
-  def prompt_invalid_input
-    puts 'Please enter a number from 1-9'
+  def prompt_invalid_space
+    puts "Please enter a valid move"
```
Comment (plopeziv) [inline lib/UI.rb]: There seem to be a few different types of error prompts that do the same job. "Please enter a valid selection between int and in" and "Please enter a valid move". I personally like the more descriptive former than the latter.

**Review by plopeziv — COMMENTED** [review]
A couple of notes perhaps not particularly related to change files. I like how descriptive your git history was. There was one that seemed a little long but it looked like the commit was trying to accomplish two things at once. I like to think of each commit to accomplish a singular goal per commit. This should help keep your descriptions short and avoid from commits bleeding into each other.

Small Readme nitpick, perhaps add that the user must bundle install your repo. I know most people probably know to do this but as a new ruby user I had to hunt the command down elsewhere. 

Lastly, maybe add a slight delay on the computer move. Computer vs Computer seemed very fast when watching a game.

Other than that good job with everything. I am constantly impressed with how much ground you cover every week!

File: lib/UI.rb (hunk 3 above)
Comment (sethpthomas91) [reply lib/UI.rb]: Got it. Fixed it up.

File: spec/ui_spec.rb (hunk 5 above)
Comment (sethpthomas91) [reply spec/ui_spec.rb:32]: I implemented this! Great suggestion.

File: spec/computer_player_spec.rb (hunk 8 above)
Comment (sethpthomas91) [reply spec/computer_player_spec.rb:26]: Agreed. Added to the list!

File: lib/UI.rb (hunk 9 above)
Comment (sethpthomas91) [reply lib/UI.rb]: Agreed. I improved this.

**Pull request merged**